File: frontend/src/api/exchange.js
import api from './client.js'

export const exchangeApi = {
  // GET /api/v1/exchange/
  // Получить списки входящих, исходящих заявок и историю обменов
  getRequests: () =>
    api.get('v1/exchange/').then(res => res.data),

  // POST /api/v1/exchange/
  // Создать новую заявку на обмен дежурствами
  // data: { initiator_duty_id: number, suggested_id: number, suggested_duty_id: number }
  createRequest: (data) =>
    api.post('v1/exchange/', data).then(res => res.data),

  // PATCH /api/v1/exchange/{exchange_id}/status
  // Изменить статус заявки (accepted, rejected, cancelled)
  updateStatus: (exchangeId, status) =>
    api.patch(`v1/exchange/${exchangeId}/status`, { status }).then(res => res.data),
}

export default exchangeApi

File: backend/app/schemas/exchange.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from app.models.exchange import ExchangeStatus
from app.schemas.group import StudentInGroup
from app.schemas.duty import DutyScheduleRead

class ExchangeCreate(BaseModel):
    initiator_duty_id: int    # ID дежурства, от которого СТУДЕНТ хочет отказаться
    suggested_duty_id: int    # ID дежурства ТОГО, с кем он хочет поменяться
    suggested_id: int         # ID студента, на чью дату мы претендуем

    model_config = ConfigDict(from_attributes=True)

class ExchangeStatusUpdate(BaseModel):
    status: ExchangeStatus    # Примет ACCEPTED, REJECTED или CANCELLED

class ExchangeResponse(BaseModel):
    id: int
    initiator_id: int
    initiator_duty_id: int
    suggested_id: int
    suggested_duty_id: int
    status: ExchangeStatus
    created_at: datetime

    initiator: StudentInGroup
    suggested: StudentInGroup
    initiator_duty: DutyScheduleRead
    suggested_duty: DutyScheduleRead

    model_config = ConfigDict(from_attributes=True)


class ExchangeListResponse(BaseModel):
  incoming: list[ExchangeResponse]  # Активные заявки текущему юзеру (PENDING)
  outgoing: list[ExchangeResponse]  # Активные заявки от текущего юзера (PENDING)
  history: list[ExchangeResponse]  # Все завершенные обмены (ACCEPTED, REJECTED, CANCELLED)

  model_config = ConfigDict(from_attributes=True)